torch/csrc/jit/script/module.h

```diff
@@ -142,70 +129,44 @@ struct TORCH_API Module {
   IValue forward(std::vector<IValue> inputs) {
     return get_method("forward")(std::move(inputs));
   }
-
   void register_buffer(const std::string& name, autograd::Variable v) {
-    if (auto b = find_attribute(name)) {
-      AT_ASSERT(b->type()->isSubtypeOf(TensorType::get()));
-      b->setValue(v);
-      return;
-    }
-    insert(
-        name,
-        attributes_,
-        EntityType::ATTRIBUTE,
-        appendSlot(name, TensorType::get(), std::move(v)));
+    get_or_add_slot(name, TensorType::get(), v, EntityType::ATTRIBUTE);
   }
 
   void register_parameter(
       const std::string& name,
       autograd::Variable v,
       bool is_buffer) {
-    if (is_buffer) {
-      register_buffer(name, std::move(v));
-      return;
-    }
-    if (auto p = find_parameter(name)) {
-      p->setValue(v);
-      return;
-    }
-    insert(
+    get_or_add_slot(
         name,
-        parameters_,
-        EntityType::PARAMETER,
-        appendSlot(name, TensorType::get(), std::move(v)));
+        TensorType::get(),
+        v,
+        is_buffer ? EntityType::ATTRIBUTE : EntityType::PARAMETER);
   }
   void register_attribute(
       const std::string& name,
       const TypePtr type,
       IValue ivalue) {
-    insert(
-        name,
-        attributes_,
-        EntityType::ATTRIBUTE,
-        appendSlot(name, type, ivalue));
+    get_or_add_slot(name, type, ivalue, EntityType::ATTRIBUTE);
   }
   void register_module(
       const std::string& name,
       std::shared_ptr<Module> module) {
-    module->field_name_ = name;
-    appendSlot(name, module->module_value_->type(), module->module_value_);
-    insert(name, modules_, EntityType::MODULE, std::move(module));
-  }
-
-  Slot parameter_slot(const std::string& name) const {
-    return parameters_[get_offset(name, EntityType::PARAMETER)];
+    get_or_add_slot(
+        name, module->type(), module->module_object(), EntityType::MODULE);
   }
 
   void set_parameter(const std::string& name, at::Tensor v) {
```

Comment:
(future PR) we have a `set_parameter` but no `set_buffer`/`set_attribute`. Probably worth adding